### ISSUE TITLE
Resource drift with catalog object tags by creating ibm_cm_object

### ISSUE THREAD

**Issue by vbontempi** [IBM-Cloud/terraform-provider-ibm#4331]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform IBM Provider Version

Terraform v1.3.7

### Affected Resource(s)

* ibm_cm_object

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
#### VPE section

module "resource_group" {
  source = "git::https://github.com/terraform-ibm-modules/terraform-ibm-resource-group.git?ref=v1.0.5"
  existing_resource_group_name = var.resource_group
}


module "sdnlb_vpe_catalog_module" {
  source               = "git::https://github.ibm.com/GoldenEye/sdnlb-module.git//sdnlb-vpe-catalog-module?ref=task_3339"
  vpe_catalog_name     = var.vpe_catalog_name
  disable_internal_vpe = var.disable_internal_vpe
  disable_service_vpe  = var.disable_service_vpe
  resource_group_id    = module.resource_group.resource_group_id
}

module "sdnlb_vpe_object" {
  source                       = "git::https://github.ibm.com/GoldenEye/sdnlb-module.git//sdnlb-vpe-object-module?ref=task_3339"
  vpe_catalog_id               = module.sdnlb_vpe_catalog_module.internal_vpe_catalog[0].id
  ibm_region                   = var.region
  vpe_object_display_name      = var.vpe_object_display_name
  vpe_object_programmatic_name = var.vpe_object_programmatic_name
  vpe_object_description       = var.vpe_object_description
  endpoint_type                = var.vpe_object_endpoint_type
  fully_qualified_domain_names = var.vpe_fqdns
  vpe_object_tags              = var.vpe_object_tags
  service_crn                  = var.vpe_object_service_crn
  dns_domain                   = var.dns_domain
}




# (...) adding code from https://github.ibm.com/GoldenEye/sdnlb-module.git//sdnlb-vpe-object-module https://github.ibm.com/GoldenEye/sdnlb-module/blob/task_3339/sdnlb-vpe-object-module/main.tf for convenience

data "ibm_cm_object" "vpe_catalog_object" {
  count      = var.existing_vpe_catalog_object_id != null ? 1 : 0
  catalog_id = var.vpe_catalog_id
  object_id  = var.existing_vpe_catalog_object_id
}

locals {
  catalog_object_data = <<-EOT
{
      "dns_domain": "${var.dns_domain}",
      "endpoint_type": "${var.endpoint_type}",
      "fully_qualified_domain_names": [${join(",", [for d in var.fully_qualified_domain_names : format("%q", d)])}],
      "service_crn": "${var.service_crn}"
}
EOT
}

resource "ibm_cm_object" "vpe_catalog_object" {
  count             = var.existing_vpe_catalog_object_id != null ? 0 : 1
  catalog_id        = var.vpe_catalog_id
  name              = var.vpe_object_programmatic_name
  label             = var.vpe_object_display_name
  kind              = var.vpe_object_kind
  short_description = var.vpe_object_description
  data              = jsonencode(local.catalog_object_data)
  parent_id         = var.ibm_region # use plain region (e.g. eu-gb) and not crn of the region
  tags              = var.vpe_object_tags
}

locals {
  vpe_catalog_object = var.existing_vpe_catalog_object_id != null ? data.ibm_cm_object.vpe_catalog_object : ibm_cm_object.vpe_catalog_object
}

```

### Debug Output

https://gist.github.com/vbontempi/e075fb83a2bddc8c8a66faac1e680f00#file-gistfile1-txt

### Expected Behavior

With a terraform apply run and a following terraform plan the catalog object shouldn't require to be updated

### Actual Behavior

The catalog object resource has 3 tags added by the provider, returned by the provider, shown by terraform show, but not present in the original terraform template

This makes terraform to update the resource at every run removing the tags that are not included in the original template

```
    ~ data              = jsonencode(
          ~ {
              - dns_domain                   = "test-provider.test.cloud.ibm.com"
              - endpoint_type                = "api"
              - fully_qualified_domain_names = [
                  - "test-vb.test-provider.test.cloud.ibm.com",
                ]
              - service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
            } -> jsonencode(
                {
                  + dns_domain                   = "test-provider.test.cloud.ibm.com"
                  + endpoint_type                = "api"
                  + fully_qualified_domain_names = [
                      + "test-vb.test-provider.test.cloud.ibm.com",
                    ]
                  + service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
                }
            )
        )
        id                = "152abe79-baad-424b-a585-91c91000c36b-test_provider_vb"
        name              = "test_provider_vb"
      ~ tags              = [
            "test-provider-vb",
          - "eu-gb",
          - "svc:service",
          - "fqdn:test-vb.test-provider.test.cloud.ibm.com",
        ]
        # (14 unchanged attributes hidden)
    }
```

### Steps to Reproduce

1. `terraform apply`

**Comment by vbontempi** [comment]
@benbuchanan the issue is still there with provider `1.51.0-beta0`

This is the apply output

```
# module.sdnlb_vpe_object.ibm_cm_object.vpe_catalog_object[0] will be created
  + resource "ibm_cm_object" "vpe_catalog_object" {
      + catalog_id        = (known after apply)
      + catalog_name      = (known after apply)
      + created           = (known after apply)
      + crn               = (known after apply)
      + data              = "\"{\\n      \\\"dns_domain\\\": \\\"test-provider.test.cloud.ibm.com\\\",\\n      \\\"endpoint_type\\\": \\\"api\\\",\\n      \\\"fully_qualified_domain_names\\\": [\\\"test-vb.test-provider.test.cloud.ibm.com\\\"],\\n      \\\"service_crn\\\": \\\"crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com\\\"\\n}\\n\""
      + id                = (known after apply)
      + kind              = "vpe"
      + label             = "provider test vb"
      + name              = "test_provider_vb"
      + object_id         = (known after apply)
      + parent_id         = "eu-gb"
      + publish           = (known after apply)
      + rev               = (known after apply)
      + short_description = "TEST provider Vb"
      + state             = (known after apply)
      + tags              = [
          + "test-provider-vb",
        ]
      + updated           = (known after apply)
      + url               = (known after apply)
    }
```

this is the plan immediately after the previous apply

```
# module.sdnlb_vpe_object.ibm_cm_object.vpe_catalog_object[0] will be updated in-place
  ~ resource "ibm_cm_object" "vpe_catalog_object" {
      ~ data              = jsonencode(
          ~ {
              - dns_domain                   = "test-provider.test.cloud.ibm.com"
              - endpoint_type                = "api"
              - fully_qualified_domain_names = [
                  - "test-vb.test-provider.test.cloud.ibm.com",
                ]
              - service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
            } -> jsonencode(
                {
                  + dns_domain                   = "test-provider.test.cloud.ibm.com"
                  + endpoint_type                = "api"
                  + fully_qualified_domain_names = [
                      + "test-vb.test-provider.test.cloud.ibm.com",
                    ]
                  + service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
                }
            )
        )
        id                = "bfa68aac-f771-433e-adab-5e6b01cb58ad-test_provider_vb"
        name              = "test_provider_vb"
      ~ tags              = [
            "test-provider-vb",
          - "eu-gb",
          - "svc:service",
          - "fqdn:test-vb.test-provider.test.cloud.ibm.com",
        ]
        # (14 unchanged attributes hidden)
    }
```

confirmed by terraform show, which shows tags auto generated into the tf object

```
# module.sdnlb_vpe_object.ibm_cm_object.vpe_catalog_object[0]:
resource "ibm_cm_object" "vpe_catalog_object" {
    catalog_id        = "bfa68aac-f771-433e-adab-5e6b01cb58ad"
    catalog_name      = "provider test vb-cp-internal-vpes"
    created           = "2023-02-27T17:11:21.362Z"
    crn               = "crn:v1:bluemix:public:globalcatalog-collection:global:a/abac0df06b644a9cabc6e44f55b3880e:bfa68aac-f771-433e-adab-5e6b01cb58ad:vpe:bfa68aac-f771-433e-adab-5e6b01cb58ad-test_provider_vb"
    data              = jsonencode(
        {
            dns_domain                   = "test-provider.test.cloud.ibm.com"
            endpoint_type                = "api"
            fully_qualified_domain_names = [
                "test-vb.test-provider.test.cloud.ibm.com",
            ]
            service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
        }
    )
    id                = "bfa68aac-f771-433e-adab-5e6b01cb58ad-test_provider_vb"
    kind              = "vpe"
    label             = "provider test vb"
    name              = "test_provider_vb"
    object_id         = "bfa68aac-f771-433e-adab-5e6b01cb58ad-test_provider_vb"
    parent_id         = "eu-gb"
    publish           = [
        {
            ibm_approved              = null
            permit_ibm_public_publish = null
            portal_approval_record    = null
            portal_url                = null
            public_approved           = null
        },
    ]
    rev               = "2-ed90c9099ddf55ce06dd598be2dc3263"
    short_description = "TEST provider Vb"
    state             = [
        {
            current           = "new"
            current_entered   = "2023-02-27T17:11:21.119Z"
            pending           = ""
            pending_requested = ""
            previous          = ""
        },
    ]
    tags              = [
        "test-provider-vb",
        "eu-gb",
        "svc:service",
        "fqdn:test-vb.test-provider.test.cloud.ibm.com",
    ]
    updated           = "2023-02-27T17:11:22.170Z"
    url               = "https://cm.globalcatalog.cloud.ibm.com/api/v1-beta/catalogs/bfa68aac-f771-433e-adab-5e6b01cb58ad/objects/bfa68aac-f771-433e-adab-5e6b01cb58ad-test_provider_vb"
}
```

**Comment by benbuchanan** [comment]
@vbontempi Not sure what happened, it looks like the PR for the tag fix didn't get merged successfully.  I'll re-add the fix and open another PR.

**Comment by benbuchanan** [comment]
@vbontempi PR for tag fix is here: https://github.com/IBM-Cloud/terraform-provider-ibm/pull/4361

Could you try something for me?  Instead of declaring the object data like this:
```
locals {
  catalog_object_data = <<-EOT
{
      "dns_domain": "${var.dns_domain}",
      "endpoint_type": "${var.endpoint_type}",
      "fully_qualified_domain_names": [${join(",", [for d in var.fully_qualified_domain_names : format("%q", d)])}],
      "service_crn": "${var.service_crn}"
}
EOT
}
```

could you try it like this?
```
locals {
  catalog_object_data = {
    endpoint_type = "${var.endpoint_type}",
    dns_domain = "${var.dns_domain}",
    fully_qualified_domain_names = [join(",", [for d in var.fully_qualified_domain_names : format("%q", d)])],
    service_crn = "${var.service_crn}"
  }
}
```

**Comment by vbontempi** [comment]
@benbuchanan is the test to be performed once the fix is merged or with the current code version?

**Comment by vbontempi** [comment]
Testing with version 1.51-beta0 it seems that this code allows to avoid the problem with the indentation. But it still produces a code that when I initially tried it caused me some concerns on the value of `fully_qualified_domain_names`:

```
resource "ibm_cm_object" "vpe_catalog_object" {
    catalog_id        = "62491a6f-77a2-45f3-b2e9-e370e2e22682"
    catalog_name      = "provider test vb-cp-internal-vpes"
    created           = "2023-02-28T10:00:27.891Z"
    crn               = "crn:v1:bluemix:public:globalcatalog-collection:global:a/abac0df06b644a9cabc6e44f55b3880e:62491a6f-77a2-45f3-b2e9-e370e2e22682:vpe:62491a6f-77a2-45f3-b2e9-e370e2e22682-test_provider_vb"
    data              = jsonencode(
        {
            dns_domain                   = "test-provider.test.cloud.ibm.com"
            endpoint_type                = "api"
            fully_qualified_domain_names = [
                "\"test-vb.test-provider.test.cloud.ibm.com\"",
            ]
            service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
        }
    )
    id                = "62491a6f-77a2-45f3-b2e9-e370e2e22682-test_provider_vb"
    kind              = "vpe"
    label             = "provider test vb"
    name              = "test_provider_vb"
    object_id         = "62491a6f-77a2-45f3-b2e9-e370e2e22682-test_provider_vb"
    parent_id         = "eu-gb"
    publish           = [
        {
            ibm_approved              = null
            permit_ibm_public_publish = null
            portal_approval_record    = null
            portal_url                = null
            public_approved           = null
        },
    ]
    rev               = "2-99e492a43fbb881d9a8c8f124f419428"
    short_description = "TEST provider Vb"
    state             = [
        {
            current           = "new"
            current_entered   = "2023-02-28T10:00:27.751Z"
            pending           = ""
            pending_requested = ""
            previous          = ""
        },
    ]
    tags              = [
        "test-provider-vb",
        "eu-gb",
        "svc:service",
    ]
    updated           = "2023-02-28T10:00:28.542Z"
    url               = "https://cm.globalcatalog.cloud.ibm.com/api/v1-beta/catalogs/62491a6f-77a2-45f3-b2e9-e370e2e22682/objects/62491a6f-77a2-45f3-b2e9-e370e2e22682-test_provider_vb"
}
```

specifically in the data field:

```
data              = jsonencode(
        {
            dns_domain                   = "test-provider.test.cloud.ibm.com"
            endpoint_type                = "api"
            fully_qualified_domain_names = [
                "\"test-vb.test-provider.test.cloud.ibm.com\"",
            ]
            service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
        }
    )
```

testing it with terraform console produces this:
```
terraform console
> jsonencode({dns_domain = "test-provider.test.cloud.ibm.com", endpoint_type = "api", fully_qualified_domain_names = ["\"test-vb.test-provider.test.cloud.ibm.com\"",],service_crn= "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"})
"{\"dns_domain\":\"test-provider.test.cloud.ibm.com\",\"endpoint_type\":\"api\",\"fully_qualified_domain_names\":[\"\\\"test-vb.test-provider.test.cloud.ibm.com\\\"\"],\"service_crn\":\"crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com\"}"
```

which is a valid JSON but I am not sure if fully_qualified_domain_names value is correct: 
```json
{
  "dns_domain": "test-provider.test.cloud.ibm.com",
  "endpoint_type": "api",
  "fully_qualified_domain_names": [
    "\"test-vb.test-provider.test.cloud.ibm.com\""
  ],
  "service_crn": "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
}
```

is the value `\"test-vb.test-provider.test.cloud.ibm.com\"` of `fully_qualified_domain_names` correctly managed from VPE point of view?

**Comment by vbontempi** [comment]
@benbuchanan I have tested your proposal but I had to make some change on my side: the fully_qualified_domain_names parameter is an array of hostnames.

I modified the input var `var.fully_qualified_domain_names` from array to string, so without leaving its management & conversion to terraform or the provider but having the module to handle its transformation to a JSON array string:
so from
```
locals {
  catalog_object_data = <<-EOT
{
      "dns_domain": "${var.dns_domain}",
      "endpoint_type": "${var.endpoint_type}",
      "fully_qualified_domain_names": [${join(",", [for d in var.fully_qualified_domain_names : format("%q", d)])}],
      "service_crn": "${var.service_crn}"
}
EOT
}
```
OR
```
OR
```
locals {
  catalog_object_data = {
    endpoint_type = "${var.endpoint_type}",
    dns_domain = "${var.dns_domain}",
    fully_qualified_domain_names = [join(",", [for d in var.fully_qualified_domain_names : format("%q", d)])],
    service_crn = "${var.service_crn}"
  }
}
```

that consider the fully_qualified_domain_names as list of string to

```
locals {
  catalog_object_data = {
    endpoint_type = "${var.endpoint_type}",
    dns_domain = "${var.dns_domain}",
    fully_qualified_domain_names = "[ ${join(",", [for s in var.fully_qualified_domain_names : format("%q", s)])} ]",
    service_crn = "${var.service_crn}"
  }
```

that formats it explicitly in a string JSON array.

By this way the indentation problem returning an update at every apply is solved.
Looking at the output of terraform when planning/applying
```
data              = jsonencode(
            {
              + dns_domain                   = "test-provider.test.cloud.ibm.com"
              + endpoint_type                = "api"
              + fully_qualified_domain_names = jsonencode(
                    [
                      + "test-vb.test-provider.test.cloud.ibm.com",
                    ]
                )
              + service_crn                  = "crn:v1:bluemix:public:service:global:::endpoint:test-provider.test.cloud.ibm.com"
            }
        )
```

it seems that the provider is now performing a `jsonencode` on the field `fully_qualified_domain_names` while in the first version the jsonencode isn't performed, so the behavior traced with this issue seems related with how the provider manages the data field according to the input type.

The net of this issue is that passing the value of fully_qualified_domain_names field as JSON string already formatted helps to solve the formatting issue, but the provider should behave in the same way if the array is passed through terraform join method which has the same result.

**Comment by benbuchanan** [comment]
@vbontempi I'm not sure I follow.  What is the problem with leaving `fully_qualified_domain_names` as an array and just passing it as is to your `catalog_object_data` like this?
```
fully_qualified_domain_names = var.fully_qualified_domain_names,
```

**Comment by vbontempi** [comment]
@benbuchanan this was the first code version, but terraform was failing due to JSON validation failure. This is why I started to switch to a different way of formatting the catalog_object_data
I tested it again, and also this problem is solved.

**Comment by vbontempi** [comment]
@benbuchanan last tests show that the provider tags aren't causing issues anymore, is it solved too?

**Comment by benbuchanan** [comment]
@vbontempi Glad to hear.  And yes it looks like the tags fix made it into the 1.51.0 provider release as well.
